roachtest/cdc: add clear error message when a golang cannot be installed

jayshrivastava opened the pull request:
### roachtest/cdc: add clear error message when a golang cannot be installed 
Previously, the test would fail immediately if installing golang, which is currently required to run the mock webhook sink, would fail. This change adds retries for up to 5 mins. If installation still fails after 5 minutes, the error is wrapped with a more clear message.

Informs: https://github.com/cockroachdb/cockroach/issues/107088
Epic: None
Release note: None

---

### roachtest: add retries to Cluster.Install
This change adds retries to `Cluster.Install` to mitigate
test failures due to package installation failures.

The webhook test in `cdc.go` is updated to use `Cluster.Install` to
install `go`.

The install command for `go` is just `sudo apt --yes install golang-go;`
now. No previously existing uses of `Cluster.Install` used this command,
so changing it should be safe.

Informs: https://github.com/cockroachdb/cockroach/issues/103316
Informs: https://github.com/cockroachdb/cockroach/issues/107088
Release note: None
Epic: None

jayshrivastava committed:
roachtest/cdc: add clear error message when a golang cannot be installed

Previously, the test would fail immediately if installing golang, which is
currently required to run the mock webhook sink, would fail. This change
adds retries for up to 5 mins. If installation still fails after 5 minutes,
the error is wrapped with a more clear message.

Informs: https://github.com/cockroachdb/cockroach/issues/107088
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107271)
<!-- Reviewable:end -->

miretskiy commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -184,7 +184,20 @@ func (ct *cdcTester) setupSink(args feedArgs) string {
+		// As seen in #107061, this can hit a 503 Service Unavailable when
+		// trying to download the package, so we retry every 30 seconds
+		// for up to 5 mins below.
+		err = retry.WithMaxAttempts(ct.ctx, retry.Options{
```
do you think it would make sense to add a helper function "InstallDeps" or something
like this which would do this logic? We also install other things, so, perhaps doing that with retries should also help.

renatolabs replied:
A good place for this would be:

https://github.com/cockroachdb/cockroach/blob/98cd2f0efedda138318b85caafa55897182ffc77/pkg/cmd/roachtest/cluster.go#L2410-L2413

renatolabs reviewed (approved):
FWIW, `apt` flakes more often than one would think (related: #103316). We're also discussing ways to define test dependencies. It will be supported at some point in the indefinite future.

jayshrivastava committed:
roachprod: add retries to Install

This change adds retries to `roachprod.Install` to mitigate
test failures due to package installation failures.

The webhook test in `cdc.go` is updated to use `Cluster.Install` to
install `go`.

The install command for `go` is just `sudo apt --yes install golang-go;`
now. No previously existing uses of `Cluster.Install` used this command,
so changing it should be safe.

Informs: https://github.com/cockroachdb/cockroach/issues/103316
Informs: #107088
Closes: https://github.com/cockroachdb/cockroach/issues/71934
Release note: None
Epic: None

jayshrivastava commented:
@renatolabs I pushed a commit adding optional retries to `(c *clusterImpl) Install`. Would you mind taking a quick look? Let me know what you think.

I also considered hardcoding retries rather than having them in an optional param, but I thought caller's might prefer to specify their own retry configuration depending on what they are installing.

renatolabs reviewed (commented):
Thanks a lot for working on this! Left some minor comments.

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-:-N_pCdEQEJcbGZdt7W0y:b-pugopo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @jayshrivastava, @miretskiy, and @srosenberg)

___
*[`-- commits` line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pAOen2WM6FmhSGq0s):*
Can we update this `roachprod: add retries to Install` (same for long description)? The commit adds retries to `roachprod` (which is good, this feature makes more sense at that level).
___
*[`-- commits` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pBSMB6h2D3YmPR8wF):*
I think it's fair for this commit to close #71934.
___
*[`pkg/roachprod/roachprod.go` line 838 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pBjV95KD2YvHO2e-T:-N_pBjV95KD2YvHO2e-U:b7igtfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca2d3b023d278057fd51adfc2bbe312d2d5f9a70/pkg/roachprod/roachprod.go#L838)):*
> ```Go
> 	}
> 
> 	// As seen in #107061, this can hit a 503 Service Unavailable when
> ```

Can this reference #103316 instead? It's the more general issue and is cross referenced with lots of instances of `apt` flakes instead of this particular instance.
___
*[`pkg/roachprod/roachprod.go` line 845 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pBtT4EgUQEigutqcJ:-N_pBtT4EgUQEigutqcK:b-2yzm88) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca2d3b023d278057fd51adfc2bbe312d2d5f9a70/pkg/roachprod/roachprod.go#L845)):*
> ```Go
> 		Multiplier:     1,
> 	}, 10, func() error {
> 		err = install.Install(ctx, l, c, software)
> ```

Nit: we can scope this `err` to this function (`err :=`). It's harmless now, but it's unexpected nonetheless.
___
*[`pkg/roachprod/roachprod.go` line 846 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pC0JHEK54aXiVHm_f:-N_pC0JHEK54aXiVHm_g:bfesqm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca2d3b023d278057fd51adfc2bbe312d2d5f9a70/pkg/roachprod/roachprod.go#L846)):*
> ```Go
> 	}, 10, func() error {
> 		err = install.Install(ctx, l, c, software)
> 		err = errors.Wrapf(err, "infrastructure failure; could not install %s", software)
> ```

Can we log this error? This will give visibility to people running `roachprod` directly and also it will also show in roachtest logs.


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
TYFR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-:-N_tQK7f214xzg44pj1n:b-80slv9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @renatolabs, and @srosenberg)

___
*[`-- commits` line 14 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pAOen2WM6FmhSGq0s):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Can we update this `roachprod: add retries to Install` (same for long description)? The commit adds retries to `roachprod` (which is good, this feature makes more sense at that level).
</blockquote></details>

Done.
___
*[`-- commits` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pBSMB6h2D3YmPR8wF):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I think it's fair for this commit to close #71934.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107271#gh-1269663432:-N_tOXmdAR8I0Wyc5eas:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad933a7b29a98de129ab13b052891c051bcf0d83/pkg/cmd/roachtest/tests/cdc.go#L190)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

A good place for this would be:

https://github.com/cockroachdb/cockroach/blob/98cd2f0efedda138318b85caafa55897182ffc77/pkg/cmd/roachtest/cluster.go#L2410-L2413
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107271#gh-1269664422:-N_tOYEf9shPCo1noMdd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad933a7b29a98de129ab13b052891c051bcf0d83/pkg/cmd/roachtest/tests/cdc.go#L199)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Maybe a better option would be to call `t.Skip()` -- your team will probably just close the issue immediately after seeing this failure anyway, so might as well not report it.
</blockquote></details>

Done.
___
*[`pkg/roachprod/roachprod.go` line 838 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pBjV95KD2YvHO2e-T:-N_tOfbb7Sl2qRnLqYI8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca2d3b023d278057fd51adfc2bbe312d2d5f9a70/pkg/roachprod/roachprod.go#L838)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Can this reference #103316 instead? It's the more general issue and is cross referenced with lots of instances of `apt` flakes instead of this particular instance.
</blockquote></details>

Done.
___
*[`pkg/roachprod/roachprod.go` line 846 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107271#-N_pC0JHEK54aXiVHm_f:-N_tP5lK-E20KLkEFmfj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca2d3b023d278057fd51adfc2bbe312d2d5f9a70/pkg/roachprod/roachprod.go#L846)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Can we log this error? This will give visibility to people running `roachprod` directly and also it will also show in roachtest logs.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jayshrivastava commented:
bors r+

renatolabs commented:
blathers backport 23.1
